Move most logic about what a Set contains into Set

billdueber opened the pull request:
This is only a reorganization -- no added or lost functionality. I wanted to tackle this before working on the deleted records and range queries.

The idea was to put as much logic about what constitutes a particular Set into the Set object itself, as read
from the `settings.yml` file. 

  * Move solr filters to settings.yml
  * Ditch the big XPath exclusion filters of disallowed rights codes for pd/pdus sets. Instead use simple lists of those codes, again put into settings.yml
  * Put the code to remove fields from a record (e.g., to ditch 974s that are ic) into Record (had been `#prune` in PartialResult
  * Move logic on how to munge records (remove 974s) and ignore records (those without 974s) into the `Set` objects. The set should now be the source of truth for what gets passed on to the next stage.
  * Fix `require` statements so individual tests can be run (mostly needed to find nokogiri and the sinatra "settings" object to create our Settings).

aelkiss commented on config/settings.yml:
```diff
@@ -6,5 +6,12 @@ admin_email: "admin@default.invalid"
+  "hathitrust:pdus":
+    name: "Public domain works according to copyright law in the United States"
+    filter_query: ["ht_searchonly:false", "ht_searchonly_intl:false"]
+    excluded_rights: [ic, op, orph, und, umall, nobody, orphcand, und-world, icus, pd-pvt, supp ]
```
The current behavior is to include *only* pdus items in the pdus set. Whether that's desired or not, I'm not sure, but this does change that behavior.

aelkiss replied:
Also, this changes the filter query: we want `ht_searchonly_intl: true` for pdus, right?

billdueber replied:
I totally screwed up the pdus filter; fixing.

I also screwed up the rights codes for `pdus`, but in my defense a feed that's _only_ `pdus` doesn't make any sense to me. As a consumer, if I want stuff that can be seen by my US-based users, I need to consume the `pd` feed (which will have none of the `pdus` 974s), then get the `pdus` feed (which has _only_ the `pdus` 974s), and then merge them somehow.  It makes sense from an item-based perspective, but not record-based.  We should find out if anyone is using it, or if there's a non-obvious use-case for it.

aelkiss commented on lib/oai_solr/record.rb:
```diff
@@ -34,6 +34,24 @@ def marc_record
+    # This is destructive on the record
+    # @yieldreturn [Boolean] Whether to keep a given field
+    # @return [OAISolr::Record] self
+    def keep_fields!(&blk)
```
Doesn't appear to be used anywhere?

billdueber replied:
Right -- put in for symmetry with `reject_fields!`

aelkiss commented on lib/oai_solr/set.rb:
```diff
@@ -50,18 +29,46 @@ def initialize
+      @excluded_rights_codes = Settings.sets[set_spec]["excluded_rights"]
+    end
+
+    def filter_record_fields(r)
```
The name is a little bit confusing in that we also have the functionality to deliver only a subset of fields (https://github.com/hathitrust/oai_solr/blob/main/lib/oai_solr/marc21.rb#L34). I wonder if (given the current needs) it would make more sense to do something like this (inlining `remove_fields!`, which is only called here, and limiting to just checking the 974)

```ruby
def filter_record_items(r)
    marc_record[974].delete_if { |f| excluded_rights_codes.include?(f["r"])
end
```

billdueber replied:
I'm not against this, but it's messier than that. It would have to be the whole thing from remove_fields!, because the part where `MARC::Record` is kind of enumerable is a messy mess. There's actually an underlying `FieldMap` that takes care of getting the fields and indexing them for faster access.  I should rewrite that code -- Ross wrote it 12 years ago.

We'd end up with something like
```ruby
def filter_record_items(rec)
  rec.marc_record.fields.each_by_tag("974") do |f| 
    rec.marc_record.fields.delete(f) if excluded_rights_codes.include?(f["r"])}
  end
end
```

...which is digging pretty deep into the implementation for `rec` (the `OAISolr::Record` object).

I was thinking the _right_ way to address this set of issues is probably to move the creation of records into the Set (`@set.add_solr_documents( @response["response"]["docs"])` and let the set give back the records, already modified and filtered as need be, but that was a pretty big API change I didn't want to mess with.

I experimented with having explicit `class HathitrustPDUS < RestrictedSet`. That doesn't make sense if we think we're going to have lots of sets (or generated-on-the-fly sets), but might if we're just going to have a handful.

billdueber replied:
Or we could just have a `set.remove_unwanted_974s(rec)` and make it specific and unambiguous.

aelkiss replied:
Let's go with `remove_unwanted_974s`

aelkiss commented:
Overall, this makes sense to me, but I'd like @moseshll  to weigh in on whether this will simplify the issues he was finding in terms of getting rights for items on the Dublin Core side.

aelkiss commented:
Need to make sure this plays well with deleted items, mainly by making sure that it doesn't try to parse nonexistent MARC.

moseshll reviewed (commented):
I'm pretty enthusiastic about this refactor, though I would like a chance to weigh in on the implementation of `remove_unwanted_974s` as discussed in the comments, before fully signing off.

aelkiss commented:
We reviewed, should be fine for #11.

Next steps:
- rebase, resolve conflicts
- make sure it deals OK with deleted items.
- change name to `remove_unwanted_974s`

billdueber commented:
Renamed that method and rebased against main. Coverage is missing the same three lines it has been (all of which are in skipped specs) -- percentage just went down because there's less code than there was before.

aelkiss commented:
OK. I turned off the coveralls status check, so we should continue to get the message without the annoying :x:

aelkiss commented:
I'm happy with this at this point. Let's merge.

aelkiss commented:
@moseshll  rebased #11  on this and it seems good. We will go ahead and merge.
